### PR TITLE
Fix #4673 - Allow setting ZoneMixing objects at Space level

### LINKED ISSUE #4673: E+ 22.2.0 - `ZoneMixing` is now allowed at space level
Allow setting ZoneMixing to a Space in model namespace?

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #4673

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/model/ThermalZone.cpp [hunk 1]
```diff
@@ -1459,12 +1459,22 @@ namespace model {
...
 
+        for (ZoneMixing& zm : space.supplyZoneMixing()) {
+          zm.hardSize();
+          zm.setParent(newSpace);
+        }
+
+        for (ZoneMixing& zm : space.exhaustZoneMixing()) {
+          // We only need to hardSize when it's the RECEIVING zone
+          zm.setSourceSpace(newSpace);
+        }
```
Comment (jmarrec) [inline src/model/ThermalZone.cpp:1476]: `ThermalZone::combineSpaces`

File: src/model/ZoneMixing.cpp [hunk 2]
```diff
@@ -270,19 +342,29 @@ namespace model {
...
+
+      if (boost::optional<double> flowRateperPerson = this->flowRateperPerson()) {
+        return this->setDesignFlowRate(*flowRateperPerson * space->numberOfPeople());
+      }
+
+      if (boost::optional<double> airChangesperHour = this->airChangesperHour()) {
+        return this->setDesignFlowRate(*airChangesperHour * space->volume() / 3600.0);
+      }
+
+      return false;
```
Comment (jmarrec) [inline src/model/ZoneMixing.cpp:367]: hardSize method

File: src/model/ZoneMixing.cpp [hunk 3]
```diff
@@ -88,45 +90,119 @@ namespace model {
...
+        LOG(Warn, "For " << briefDescription() << ", Source Space cannot be the same as the Receiving Space '" << space.nameString() << "'.");
+        return false;
+      }
+      if (receiving.iddObjectType() != IddObjectType::OS_Space) {
+        LOG(Warn, "For " << briefDescription() << ", Receiving is a Zone, so you cannot set Source as a Space.");
+        return false;
+      }
+
+      return setPointer(OS_ZoneMixingFields::SourceZoneorSpaceName, space.handle());
+    }
```
Comment (jmarrec) [inline src/model/ZoneMixing.cpp:199]: Protect users from themselves

File: src/model/ZoneMixing.hpp [hunk 4]
```diff
@@ -54,18 +57,23 @@ namespace model {
...
+    bool hardSize();
+
     /** @name Getters */
     //@{
 
     /// Returns the ThermalZone which receives air from this mixing object.
-    ThermalZone zone() const;
+    boost::optional<ThermalZone> zone() const;
+    boost::optional<Space> space() const;
+    ModelObject zoneOrSpace() const;
```
Comment (jmarrec) [inline src/model/ZoneMixing.hpp:76]: this is the API break

File: src/model/ZoneMixing.hpp [hunk 5]
```diff
@@ -74,14 +82,17 @@ namespace model {
 
     boost::optional<double> designFlowRate() const;
 
-    boost::optional<double> flowRateperZoneFloorArea() const;
+    OS_DEPRECATED boost::optional<double> flowRateperZoneFloorArea() const;
+    boost::optional<double> flowRateperFloorArea() const;
```
Comment (jmarrec) [inline src/model/ZoneMixing.hpp:86]: Lots of deprecations

File: src/model/Space.hpp [hunk 6]
```diff
@@ -625,6 +626,15 @@ namespace model {
...
 
+    /** Returns all ZoneMixing objects associated with this space, includes supply and exhaust mixing objects */
+    std::vector<ZoneMixing> zoneMixing() const;
+
+    /** Returns all ZoneMixing objects which supply air to this space */
+    std::vector<ZoneMixing> supplyZoneMixing() const;
+
+    /** Returns all ZoneMixing objects which exhaust air from this space */
+    std::vector<ZoneMixing> exhaustZoneMixing() const;
+
```
Comment (jmarrec) [inline src/model/Space.hpp:637]: Space gets some helpers

File: src/model/test/ZoneMixing_GTest.cpp [hunk 7]
```diff
@@ -205,3 +214,268 @@ TEST_F(ModelFixture, ZoneMixing_ZoneRemove) {
...
+    EXPECT_FALSE(zm.sourceSpace());
+
+    EXPECT_TRUE(zm.setSourceZone(z2));
+    ASSERT_TRUE(zm.sourceZoneOrSpace());
+    EXPECT_EQ(z2, zm.sourceZoneOrSpace().get());
+    ASSERT_TRUE(zm.sourceZone());
+    EXPECT_EQ(z2, zm.sourceZone().get());
+    EXPECT_FALSE(zm.sourceSpace());
+  }
+}
```
Comment (jmarrec) [inline src/model/test/ZoneMixing_GTest.cpp:286]: Test protecting users

File: src/model/test/ZoneMixing_GTest.cpp [hunk 8]
```diff
@@ -205,3 +214,268 @@ TEST_F(ModelFixture, ZoneMixing_ZoneRemove) {
...
+  EXPECT_FALSE(zmACH.flowRateperFloorArea());
+  EXPECT_FALSE(zmACH.flowRateperPerson());
+  EXPECT_FALSE(zmACH.airChangesperHour());
+  ASSERT_TRUE(zmACH.space());
+  EXPECT_EQ(space12_.get(), zmACH.space().get());
+  EXPECT_FALSE(zmACH.zone());
+  ASSERT_TRUE(zmACH.sourceSpace());
+  EXPECT_EQ(space34_.get(), zmACH.sourceSpace().get());
+  EXPECT_FALSE(zmACH.sourceZone());
+}
```
Comment (jmarrec) [inline src/model/test/ZoneMixing_GTest.cpp:481]: Detail ThermalZone::combineSpaces test, and the order in which things are done. Some of that is tested in FT, but not all. So I think I like having both tests

File: src/energyplus/ForwardTranslator/ForwardTranslateZoneMixing.cpp [hunk 9]
```diff
@@ -50,21 +52,64 @@ namespace openstudio {
...
+      // Let's play it safe though
+      LOG(Warn, modelObject.briefDescription() << " has the same Receiving and Source Zone or Space, it will not be translated.");
+      if (!m_excludeSpaceTranslation) {
+        // We don't allow this at model time, the only reason we expect this to happen is when m_excludeSpaceTranslation is true, we call
+        // combineSpaces, and if the user has a ZoneMixing pointing to two spaces from the same ThermalZone, you end up with matching Receiving and
+        // Source Spaces
+        OS_ASSERT(false);
+      }
+      return boost::none;
+    }
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslateZoneMixing.cpp:97]: Ft adjust + warnings in case the user had a ZoneMixing pointing to two spaces **inside the same zone** AND **opting out of space translation**. Yep, that's the state of all the corner cases we have to test for nowadays...

File: src/energyplus/Test/ZoneMixing_GTest.cpp [hunk 10]
```diff
@@ -0,0 +1,500 @@
...
+        EXPECT_EQ(equivm3s_ACH, zm.getDouble(ZoneMixingFields::DesignFlowRate).get());
+        EXPECT_TRUE(zm.isEmpty(ZoneMixingFields::FlowRateperFloorArea));
+        EXPECT_TRUE(zm.isEmpty(ZoneMixingFields::FlowRateperPerson));
+        EXPECT_TRUE(zm.isEmpty(ZoneMixingFields::AirChangesperHour));
+        EXPECT_EQ(z2.nameString(), source_->nameString());
+        EXPECT_TRUE(zm.isEmpty(ZoneMixingFields::DeltaTemperature));
+      }
+    }
+  }
+}
```
Comment (jmarrec) [inline src/energyplus/Test/ZoneMixing_GTest.cpp:388]: FT Tests when different zone.
* Test all combinations of Spaces one
* Test one zone one
* Do that when excluding spaces and when including them

File: src/energyplus/Test/ZoneMixing_GTest.cpp [hunk 11]
```diff
@@ -0,0 +1,500 @@
...
+    // EXPECT_EQ(1, ft.warnings().size());
+    EXPECT_GE(ft.warnings().size(), 1);
+    std::string expectedWarning =
+      "Object of type 'OS:ZoneMixing' and named 'zmAbs' doesn't have a Source Zone or Space, it will not be translated. As you were using "
+      "Space-Level ZoneMixing, and you are not translating to Spaces, it's possible it was pointing to two spaces inside the same zone";
+    auto warnings = ft.warnings();
+    EXPECT_NE(warnings.cend(),
+              std::find_if(warnings.cbegin(), warnings.cend(), [&expectedWarning](auto& l) { return l.logMessage() == expectedWarning; }));
+  }
+}
```
Comment (jmarrec) [inline src/energyplus/Test/ZoneMixing_GTest.cpp:500]: Make a Space Level ZoneMixing that points to two spaces **inside the same zone**:
* When trnaslating to spaces: it should work fine
* When excluding spaces: it shouldn't translate it and we should get a nice warning.

Note that I noticed we try to translate it 4 times and opened https://github.com/NREL/OpenStudio/issues/4704

**Comment by shorowit** [comment]
- [x] Verified that I get the same behavior as the previous OS 3.5 install (i.e., an EnergyPlus EMS error) when ZoneMixing is assigned to thermal zones
- [x] Verified that the EnergyPlus error goes away when ZoneMixing is assigned to spaces

I haven't tried a comprehensive set of scenarios but it works for my case. Thanks @jmarrec!
